File: ERP/Areas/Comercial/Controllers/ListaPreciosController.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;
using ENTIDADES.comercial;
using ERP.Controllers;
using ERP.Models.Ayudas;
using INFRAESTRUCTURA.Areas.Comercial.DAO;
using INFRAESTRUCTURA.Areas.Comercial.Interfaz;
using INFRAESTRUCTURA.Areas.Comercial.listaprecios;
using ENTIDADES.Identity;
using Erp.Persistencia.Modelos;

using Erp.Persistencia.Servicios;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using Erp.Infraestructura.Areas.Comercial.listaprecios.busqueda;
using Erp.Infraestructura.Areas.Ventas.DAO;

namespace ERP.Areas.Comercial.Controllers
{
    [Authorize]
    [Area("Comercial")]
    [ResponseCache(NoStore = true, Duration = 0)]
    public class ListaPreciosController : _baseController
    {
        private readonly IListaPreciosEF EF;
        private readonly Modelo db;
        private readonly ListaPreciosDAO DAO; 
        private readonly IWebHostEnvironment ruta;
        private readonly CanalVentaDAO dao_cv;

        public ListaPreciosController(IWebHostEnvironment _ruta,Modelo db_,IListaPreciosEF _EF, ICryptografhy cryptografhy, SignInManager<AppUser> signInManager) : base(cryptografhy, signInManager)
        {
            EF = _EF;
            db = db_;
            ruta = _ruta;
            LeerJson settings = new LeerJson();
            DAO = new ListaPreciosDAO(settings.GetConnectionString());
            dao_cv = new CanalVentaDAO(settings.GetConnectionString());
        }
        [Authorize(Roles = ("ADMINISTRADOR, VER LISTA DE PRECIOS"))]
        public IActionResult Index()
        {
         
            datosinicio();
            return View(EF.ListarListas());
        }
        [Authorize(Roles = ("ADMINISTRADOR, VER LISTA DE PRECIOS"))]
        public IActionResult EditarPrecios(int idlista)
        {
            datosinicio();
            var aux = db.LISTAPRECIOS.Find(idlista);
            if (aux is null)
                return NotFound();
            return View(aux);
        }
        [Authorize(Roles = ("ADMINISTRADOR, EDITAR LISTA DE PRECIOS"))]
        public IActionResult SubirLista()
        {
            datosinicio();
            return View(EF.ListarListas());
        }
        [Authorize(Roles = ("ADMINISTRADOR, EDITAR LISTA DE PRECIOS"))]
        [HttpPost]
        public IActionResult SubirLista(string lista, bool codadesy)
        {
            var data = JsonConvert.DeserializeObject<List<PreciosProducto>>(lista);
            var res = DAO.RegistrarPrecio(data, getIdEmpleado().ToString(), codadesy);
            return Json(res);
        }
        [Authorize(Roles = ("ADMINISTRADOR, CREAR LISTA DE PRECIOS"))]
        [HttpPost]
        public async Task<IActionResult> RegistrarEditar(ListaPrecios obj)
        {
            return Json(await EF.RegistrarEditarAsync(obj));
        } 
        public IActionResult ListarListas( )
        {
            return Json(EF.ListarListas());
        }
        public IActionResult ListarCanalesPrecios()
        {
            return Json(dao_cv.ListarCanalesPrecios());
        }
        public  IActionResult ListarListasHabilitadas( )
        {
            return Json( EF.ListarListasHabilitadas());
        }
        public async Task<IActionResult> BuscarProductosListaPrecios(BuscarProductosLista.Ejecutar obj)
        {
            return Json(await _mediator.Send(obj));
        }
        //
        public async Task<IActionResult> BuscarProductosListaConIncentivo(BuscarProductosListaConIncentivo.Ejecutar obj)
        {
            //EARTCOD1009 <--
            if (obj.sucursal is 0)
                obj.sucursal = getIdSucursal();
            //TEMPORAL
            var mediador = _mediator.Send(obj);
            var json_ = Json(await mediador);
            return Json(await mediador);
        }

        public async Task<IActionResult> DescargarLista(int top, int lista, string producto, string laboratorio,string tipo)
        {
            return Json(await DAO.GenerarExcelListaAsync(top,lista,producto,laboratorio,tipo, ruta.WebRootPath));
        }
        [Authorize(Roles = ("ADMINISTRADOR, EDITAR LISTA DE PRECIOS"))]
        [HttpPost]
            

        public async Task<IActionResult> EditarPreciosProducto(List<PreciosProducto> lista)
        {       
            return Json(await EF.EditarPreciosProducto(lista));
        }
        public IActionResult ListarPreciosSucursal(int? idsucursal)
        {
            if (idsucursal is null)
                idsucursal = getIdSucursal();
            return Json( EF.ListarPreciosSucursal(idsucursal.Value));
        }
        public IActionResult DuplicarPreciosLista(int lista1, int lista2)
        {

            return Json(DAO.DuplicarPreciosLista(lista1, lista2,getIdEmpleado().ToString()));
        }
        public IActionResult ListarListaXProducto(int idproducto)
        {

            return Json(EF.ListarListaXProducto(idproducto));
        }
        public IActionResult ListarPreciosxListasyProducto(string listas, int idproducto)
        {
            return Json(EF.ListarPreciosxListasyProducto(listas,idproducto));
        }
        public async Task<IActionResult> BuscarProductoByItemLista(BuscarProductoByItemLista.Ejecutar obj)
        {
            return Json(await _mediator.Send(obj));
        }
        public async Task<IActionResult> ActualizarPrecios(List<string []> arreglo) {

            return Json(DAO.ActualizarPrecios(arreglo));
          
        }
    }
}

File: ERP/wwwroot/js/controller/comercial/ListaPreciosController.js
class ListaPreciosController {
    RegistrarEditar(obj, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/RegistrarEditar";      
        $.post(url, obj).done(function (data) {
            if (data.mensaje === 'ok') 
            {
                mensaje('S', 'Datos guardados', 'TC');
                fn(data);
            }
             else 
                mensaje('W', data.mensaje, 'TC');
             
            
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    ListarListas( fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/ListarListas";      
        $.post(url).done(function (data) {           
                fn(data);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }

    ListarCanalesPrecios(fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/ListarCanalesPrecios";
        $.post(url).done(function (data) {
            fn(data);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }

    ListarListasHabilitadas(fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/ListarListasHabilitadas";      
        $.post(url).done(function (data) {           
                fn(data);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    ListarPreciosxListasyProducto(obj,fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/ListarPreciosxListasyProducto";      
        $.post(url,obj).done(function (data) {           
                fn(data);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    BuscarProductosListaPrecios(obj,fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/BuscarProductosListaPrecios";      
        $.post(url, obj).done(function (data) {              
            fn((data));
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    BuscarProductosListaConIncentivo(obj, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/BuscarProductosListaConIncentivo";      
        $.post(url, obj).done(function (data) {  
            //console.log("este es el controlador");//TEMPORAL
            //console.log(data);//TEMPORAL
            fn((data));
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
        
    }
    EditarPreciosProducto(obj, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/EditarPreciosProducto";  
        console.log(obj);
        $.post(url,obj).done(function (data) {           
            if (data.mensaje === 'ok') {
                fn(data);
                alertaSwall('S', 'Se actualizo el precio de ' + obj.lista.length + ' productos', '');
            }              
            else
                mensaje('W', data.mensaje);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    DuplicarPreciosLista(obj, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/DuplicarPreciosLista";        
        $.post(url,obj).done(function (data) {           
            if (data === 'ok')             
                alertaSwall('S', 'Precios copiados', '');                          
            else
                mensaje('W', data);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    ListarListaXProducto(idproducto, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/ListarListaXProducto?idproducto=" + idproducto;        
        $.post(url).done(function (data) {           
            fn(data);
        }).fail(function (data) {
            fn(null);
            mensajeError(data);
        });
    }
    BuscarProductoByItemLista(idprecio, fn,fnerror) {
        var url = ORIGEN + "/Comercial/ListaPrecios/BuscarProductoByItemLista?idprecio=" + idprecio;        
        $.post(url).done(function (data) {       
            if (data.mensaje == 'ok')
                fn(data.objeto);
            else { mensaje('W', data.mensaje); fnerror(); }
        }).fail(function (data) {
            fnerror();
            mensajeError(data);
        });
    }
    ExportarLista(obj, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/DescargarLista";
        $.post(url, obj).done(function (data) {
            if (data.mensaje === 'ok') {
                location.href = ORIGEN + data.objeto;
            }
            else
                mensaje('W', data.mensaje);
        }).fail(function (data) {
            mensajeError(data);
        });
    }
    SubirLista(obj, fn,fneeror) {
        var url = ORIGEN + "/Comercial/ListaPrecios/SubirLista";
        $.post(url, obj).done(function (data) {
            if (data.mensaje == 'ok') {
                mensaje('S', 'Datos guardados');
                fn(data);
            } else {
                mensaje('W', data.mensaje);
                fneeror();
            }
        }).fail(function (data) {
            mensajeError(data);
            fneeror();
        });
    }
    ActualizarPrecios(obj, fn) {
        var url = ORIGEN + "/Comercial/ListaPrecios/ActualizarPrecios";
        console.log(obj);
        $.post(url, obj).done(function (data) {
            if (data === 'ok') {
                //fn(data);
                alertaSwall('S', 'Se actualizo el precio de ' + obj.lista.length + ' productos', '');
            }
            else
                mensaje('W', data.mensaje);
        }).fail(function (data) {
            //fn(null);
            mensajeError(data);
        });
    }
}